pinot-controller/src/main/java/com/linkedin/pinot/controller/api/restlet/resources/LLCSegmentCommitUpload.java

```diff
@@ -0,0 +1,148 @@
+    final String offsetStr = getReference().getQueryAsForm().getValues(SegmentCompletionProtocol.PARAM_OFFSET);
+    final String segmentName = getReference().getQueryAsForm().getValues(SegmentCompletionProtocol.PARAM_SEGMENT_NAME);
+    final String instanceId = getReference().getQueryAsForm().getValues(SegmentCompletionProtocol.PARAM_INSTANCE_ID);
+
+    if (offsetStr == null || segmentName == null || instanceId == null) {
+      LOGGER.error("Invalid call: offset={}, segmentName={}, instanceId={}", offsetStr, segmentName, instanceId);
+      return false;
+    }
+    _segmentNameStr = segmentName;
+    _instanceId = instanceId;
+    try {
+      _offset = Long.valueOf(offsetStr);
+    } catch (NumberFormatException e) {
+      LOGGER.error("Invalid offset {} for segment {} from instance {}", offsetStr, segmentName, instanceId);
+      return false;
+    }
+    return true;
+  }
+
+  boolean uploadSegment(final String instanceId, final String segmentNameStr) {
+    // 1/ Create a factory for disk-based file items
+    final DiskFileItemFactory factory = new DiskFileItemFactory();
+
+    // 2/ Create a new file upload handler based on the Restlet
+    // FileUpload extension that will parse Restlet requests and
+    // generates FileItems.
+    final RestletFileUpload upload = new RestletFileUpload(factory);
+    final List<FileItem> items;
+
+    try {
+      // The following statement blocks until the entire segment is read into memory/disk.
+      items = upload.parseRequest(getRequest());
+
+      File dataFile = null;
+
+      // TODO: refactor this part into a util method (almost duplicate code in PinotSegmentUploadRestletResource and
+      // PinotSchemaRestletResource)
+      for (FileItem fileItem : items) {
+        String fieldName = fileItem.getFieldName();
+        if (dataFile == null) {
+          if (fieldName != null && fieldName.equals(segmentNameStr)) {
+            dataFile = new File(tempDir, fieldName);
+            fileItem.write(dataFile);
+          } else {
+            LOGGER.warn("Invalid field name: {}", fieldName);
+          }
+        } else {
+          LOGGER.warn("Got extra file item while uploading LLC segments: {}", fieldName);
+        }
+
+        // Remove the temp file
+        // When the file is copied to instead of renamed to the new file, the temp file might be left in the dir
+        fileItem.delete();
+      }
+
+      if (dataFile == null) {
+        LOGGER.error("Segment not included in request. Instance {}, segment {}", instanceId, segmentNameStr);
+        return false;
+      }
+      // We will not check for quota here. Instead, committed segments will count towards the quota of a
+      // table
+      LLCSegmentName segmentName = new LLCSegmentName(segmentNameStr);
+      final String rawTableName = segmentName.getTableName();
+      final File tableDir = new File(baseDataDir, rawTableName);
+      final File segmentFile = new File(tableDir, segmentNameStr + _offset + _instanceId);
```

Comment:
implement a method to create the file name. For now, put it in the SegmentCompletionProtocol file. with a TODO to find a place for it